cfy_manager/components/postgresql_server/postgresql_server.py

```diff
@@ -1732,15 +1739,90 @@ def start(self):
             service.verify_alive(POSTGRES_SERVICE_NAME)
         logger.notice('PostgreSQL Server successfully started')
 
+    def _save_encoding_and_locale(self):
+        self.encoding = subprocess.run(
+            ['psql', '-U', 'postgres', '-c', 'SHOW SERVER_ENCODING', '-t'],
+            stdout=subprocess.PIPE, stderr=subprocess.DEVNULL
+        ).stdout.decode('utf-8').strip()
+
+        self.locale = subprocess.run(
+            ['psql', '-U', 'postgres', '-c', 'SHOW LC_CTYPE', '-t'],
+            stdout=subprocess.PIPE, stderr=subprocess.DEVNULL
+        ).stdout.decode('utf-8').strip()
+
     def stop(self, force=True):
+        self._save_encoding_and_locale()
         logger.notice('Stopping PostgreSQL Server...')
         if config[POSTGRESQL_SERVER]['cluster']['nodes']:
             service.stop('etcd')
             service.stop('patroni')
         else:
-            service.stop(POSTGRES_SERVICE_NAME)
+            try:
+                service.stop(POSTGRES_SERVICE_NAME)
+                self.upgrading_psql = False
+            except ProcessExecutionError:
+                service.stop(OLD_POSTGRES_SERVICE_NAME)
+                service.remove(OLD_POSTGRES_SERVICE_NAME)
+                self.upgrading_psql = True
         logger.notice('PostgreSQL Server successfully stopped')
 
+    def upgrade(self):
+        if self.upgrading_psql:
+            logger.notice("Upgrading PostgreSQL database version...")
+            logger.debug('Configuring and initializing new PostgreSQL '
+                         'service...')
+            self._configure_postgresql_server_service()
+            service.reload(POSTGRES_SERVICE_NAME)
+
+            # wait for postgresql service to exit after reload,
+            # since initdb needs it down.
+            # The service stops since "/var/lib/pgsql/14/data/postgresql.conf"
+            #   doesn't exist yet
+            self._verify_postgres_stopped()
+
+            res = common.run(
+                ['sudo', '-u', 'postgres',
+                 join(PGSQL_USR_DIR, 'bin', 'initdb'),
+                 '-D', PGSQL_DATA_DIR,
+                 '--locale', self.locale, '-E', self.encoding],
+                stdout=subprocess.PIPE, stderr=subprocess.PIPE,
+                ignore_failures=True
+            )
+            if 'Success' not in res.aggr_stdout:
+                raise ProcessExecutionError(
+                    f"Error initializing {POSTGRES_SERVICE_NAME} "
+                    f"database:\n{res.aggr_stdout}\n{res.aggr_stderr}")
+            logger.debug(f'Success initializing '
+                         f'{POSTGRES_SERVICE_NAME} database!')
```

Comment:
I think it'd be better if we used `self._init_postgresql_server()` here. It has more things than just the initdb, too, like setting the log dir, and some chowns (which this code forgot about)